extensions/amp-document-recommendations/0.1/amp-document-recommendations.js

```diff
@@ -229,6 +254,48 @@ export class AmpDocumentRecommendations extends AMP.BaseElement {
     this.appendNextArticle_();
     return Promise.resolve();
   }
+
+  /**
+   * Handles updates from the {@code PositionObserver} indicating a change in
+   * the position of a recommendation unit in the viewport.
+   * @param {number} i Index of the documentRef this recommendation unit is
+   *     attached to.
+   * @param
+   * {!../../../src/service/position-observer/position-observer-worker.PositionInViewportEntryDef}
+   *     position Position of the current recommendation unit in the viewport.
+   */
+  positionUpdate_(i, position) {
+    // We're only interested when the recommendations exit the viewport
+    if (position.positionRect !== null) {
+      return;
+    }
+
+    if (position.relativePos === 'top') {
+      this.setActiveDocument_(this.documentRefs_[i + 1]);
+    } else if (position.relativePos === 'bottom') {
+      this.setActiveDocument_(this.documentRefs_[i]);
+    }
+  }
+
+  /**
+   * Sets the specified document as active, updating the document title and URL.
+   * @param {!DocumentRef} documentRef Reference to the document to set as
+   *     active.
+   */
+  setActiveDocument_(documentRef) {
+    const amp = documentRef.amp;
+    this.win.document.title = amp.title || '';
+    if (this.win.history.replaceState) {
```

Comment:
AMP has a service wrapper for the history API in [src/service/history-impl.js](https://github.com/ampproject/amphtml/blob/master/src/service/history-impl.js) that is exposed in [src/services.js#historyForDoc](https://github.com/ampproject/amphtml/blob/068f00395bf9f677c2429e0b288e8ce18f09bdae/src/services.js#L221)

Reply:
This updates the URL shown, but I believe it breaks the back button for the user. Are you able to use `push`/`pop` instead?

Reply:
Hmm ok I see how that could be confusing for the user. Have you checked how `replaceState` would behave when the document is rendered in the viewer? You can [upload to Heroku](https://docs.google.com/document/d/1KSvv1oqEGKMcqKJ-aft9Buf6Be2xs3CIkyoeFa0Wqmw/edit?usp=sharing) and use the viewer test tool at https://search.google.com/test/amp

Reply:
When rendered in the viewer the replaceState method is replaced with the one from history-impl, which will send a message to the viewer instead of updating local history. I've got a TODO there to send a request to the viewer with the new URL/title which will either be a whole new message (so I won't use replaceState at all in that case), or I'll update the viewer API to add support.